cunumeric/module.py

```diff
@@ -3279,6 +3279,87 @@ def diagonal(
     )
 
 
+@add_boilerplate("a", "val")
+def fill_diagonal(a: ndarray, val: ndarray, wrap: bool = False) -> None:
+    """
+    Fill the main diagonal of the given array of any dimensionality.
+
+    For an array a with a.ndim >= 2, the diagonal is the list of locations with
+    indices a[i, ..., i] all identical. This function modifies the input
+    array in-place, it does not return a value.
+
+    Parameters
+    ----------
+
+    a : array, at least 2-D.
+        Array whose diagonal is to be filled, it gets modified in-place.
+    val : scalar or array_like
+        Value(s) to write on the diagonal. If val is scalar, the value is
+        written along the diagonal.
+        If array-like, the flattened val is written along
+        the diagonal, repeating if necessary to fill all diagonal entries.
+    wrap : bool
+        If true, the diagonal "wraps" after N columns, for tall 2d matrices.
+
+    Raises
+    ------
+    ValueError
+        If the dimension of `a` is less than 2.
+
+    Notes
+    -----
+
+    See Also
+    --------
+    numpy.fill_diagonal
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    if val.size == 0 or a.size == 0:
+        return
+
+    if a.ndim < 2:
+        raise ValueError("array must be at least 2-d")
+
+    n = _builtin_min(a.shape)
+
+    if a.ndim > 2:
+        for s in a.shape:
+            if s != n:
+                raise ValueError(
+                    "All dimensions of input must be of equal length"
+                )
+
+        if wrap:
+            raise ValueError("wrap can be set to True only for 2d matrices")
```

Comment:
It looks like NumPy allows `wrap=True` even for `Nd` matrices (of course it doesn't do anything, since those must be square), so I would remove this.